packages/core-data/src/actions.js

```diff
@@ -139,6 +155,11 @@ export function receiveEmbedPreview( url, preview ) {
 	};
 }
 
+export function* deleteEntityRecord( kind, name, recordId ) {
+	const edits = { toDelete: true };
+	yield editEntityRecord( kind, name, recordId, edits );
```

Comment:
Is there a reason we're implementing `deleteEntityRecord` using `editEntityRecord`? It makes the already-complicated `editEntityRecord` action more complicated, and potentially will cause issues if someone tries to use `@wordpress/data` with an entity that contains a `toDelete` property 🙂

I was expecting `deleteEntityRecord` to look something like this snippet which I've spliced together from different parts of `editEntityRecord` and have not tested:

```js
export function* deleteEntityRecord( kind, name, recordId ) {
	const entities = yield getKindEntities( kind );
	const entity = find( entities, { kind, name } );
	if ( ! entity ) {
		return;
	}

	yield {
		type: 'DELETE_ENTITY_RECORD_START',
		kind,
		name,
		recordId,
	};

	const persistedRecord = yield select(
		'getEntityRecord',
		kind,
		name,
		recordId
	);

	yield removeItems( kind, name, recordId );

	let error;

	try {
		let path = `${ entity.baseURL }/${ recordId }`;

		if ( entity.forceDelete ) {
			path += '?force=true';
		}

		yield apiFetch( {
			path,
			method: 'DELETE',
		} );
	} catch ( _error ) {
		error = _error;

		yield receiveEntityRecords(
			kind,
			name,
			persistedRecord,
			undefined,
			true
		);
	}

	yield {
		type: 'DELETE_ENTITY_RECORD_FINISH',
		kind,
		name,
		recordId,
		error,
	};
}
```

Reply:
My not so great reason is that I *think* that we lose undo if we go around edit. And also because my logic is that delete is just an edit until you save. In our case we're saving immediately but in other cases we would not.

I also think that the correct implementation would be to have in edit-navigation an entity provider which would hook the action that removes the block to a delete entity call and then on save we actually perform the DELETE call.

Reply:
There's some intricacies here and so, first of all, thank you for forcing me to do some research here and think through them 😀

Dispatching `undo()` and `redo()` only affects the data stored in an entity's `edits` state. In other words, only changes made to an entity using `editEntityRecord()` (viewable using `getEditedEntityRecord()`) are changed when one dispatches `undo()` or `redo()`.

`saveEntityRecord()`, `getEntityRecord()` and `deleteEntityRecord()` operate on a different level to `undo()`, `redo()` and `editEntityRecord()`. The former actions are concerned with loading, saving and deleting the entity as a whole whereas the latter actions are concerned only with the entity's `edits` state.

I wouldn't expect that one would be able to undo the changes made by `deleteEntityRecord()` just as I wouldn't expect one to be able to undo the changes made by `saveEntityRecord()` or by (the fetching that occurs in) `getEntityRecord()`.

You're right that we will want to one day implement undo and redo in `edit-navigation`. To do this we will likely stop storing the `core/navigation` and `core/navigation-link` blocks in React state as we do now and instead store them in the `edits` of a `post` entity that is never saved. That will let us use `undo()` and `redo()` to undo and redo e.g. removing a `core/navigation-link` block. When the _Save Navigation_ button is clicked, the blocks in the unsaved `post` entity will be translated into `saveEntityRecord( 'menu-item', ... )` and `deleteEntityRecord( 'menu-item', ... )` calls.

> on save we actually perform the DELETE call.

This already happens. `menu-item`s are not created or deleted until _Save Navigation_ is clicked which in turn calls `saveBlocks()` in `useNavigationBlocks()`.